File: 2023/day1/solution.py
def part1():
	result = 0
	d1 = 0
	d2 = 0

	with open("input.txt", 'r') as file:
		for line in file:
			for c in line:
				if c.isdigit():
					d1 = int(c)
					break
			for c in reversed(line):
				if c.isdigit():
					d2 = int(c)
					break
			result += d1 * 10 + d2

	print(f"Part 1 answer is: {result}")

def part2():
	result = 0
	d1 = 0
	d2 = 0
	numbers = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]

	with open("input.txt", 'r') as file:
		for line in file:
			d1 = 0
			for i, char in enumerate(line):
				if d1 == 0 and char.isdigit():
					d1 = int(char)
				for index, number in enumerate(numbers):
					if d1 == 0 and line[i:].find(number) == 0:
						d1 = index + 1
			d2 = 0
			line = ''.join(reversed(line))
			for i, char in enumerate(line):
				if d2 == 0 and char.isdigit():
					d2 = int(char)
				for index, number in enumerate(numbers):
					if d2 == 0 and line[i:].find(number[::-1]) == 0:
						d2 = index + 1
			result += d1 * 10 + d2

	print(f"Part 2 answer is: {result}")


if __name__ == '__main__':
	part1()
	part2()

File: 2023/day1/solution.rs
fn main() {
    let file_content = include_str!("input.txt");

    // Part 1
    let result_p1 = file_content.lines().fold(0, |acc, line| {
        let first = line.chars().find_map(|c| c.to_digit(10)).unwrap_or(0);
        let last = line.chars().rev().find_map(|c| c.to_digit(10)).unwrap_or(0);
        acc + first * 10 + last
    });

    println!("Part 1 answer is: {result_p1}");

    // Part 2
    let numbers = vec![
        "one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "1", "2", "3", "4",
        "5", "6", "7", "8", "9",
    ];

    let find_number = |slice: &str| {
        numbers.iter().enumerate().find_map(|(index, number)| {
            if slice.starts_with(number) {
                Some((index as u32 % 9) + 1)
            } else {
                None
            }
        })
    };

    let result_p2 = file_content.lines().fold(0, |acc, line| {
        let first = line
            .char_indices()
            .find_map(|(i, _)| find_number(&line[i..]))
            .unwrap_or(0);
        let last = line
            .char_indices()
            .rev()
            .find_map(|(i, _)| find_number(&line[i..]))
            .unwrap_or(0);
        acc + first * 10 + last
    });

    println!("Part 1 answer is: {result_p2}");
}
